### PR TITLE
[SPARK-20750][SQL] Built-in SQL Function Support - REPLACE

### PR DESCRIPTION
**Pull request by kiszk**
## What changes were proposed in this pull request?

This PR adds built-in SQL function `(REPLACE(<string_expression>, <search_string> [, <replacement_string>])`

`REPLACE()` return that string that is replaced all occurrences with given string.

## How was this patch tested?

added new test suites

### CODE REVIEW

**Comment by kiszk** [comment]
Jenkins, retest this please

**Comment by kiszk** [comment]
Jenkins, retest this please

File: conf/docker.properties.template [hunk 1]
```diff
@@ -15,6 +15,6 @@
 # limitations under the License.
 #
 
-spark.mesos.executor.docker.image: <image built from `../docker/spark-mesos/Dockerfile`>
+spark.mesos.executor.docker.image: <image built from `../external/docker/spark-mesos/Dockerfile`>
```
Comment (kiszk) [inline conf/docker.properties.template]: Thank you very much. Absolutely, this is not related.

**Comment by kiszk** [comment]
ping @gatorsmile

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 2]
```diff
@@ -835,6 +835,12 @@ public static UTF8String concatWs(UTF8String separator, UTF8String... inputs) {
     return res;
   }
 
+  public UTF8String replace(UTF8String search, UTF8String replace) {
+    String replaced = toString().replace(
+      search.toString(), replace.toString());
```
Comment (gatorsmile) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: When the search string is an empty string, we should not change the original string.

Comment (kiszk) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Good catch. Done. Test case was also added.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 3]
```diff
@@ -340,6 +340,42 @@ case class EndsWith(left: Expression, right: Expression) extends StringPredicate
   }
 }
 
+/**
+ * Replace all occurrences with string.
+ */
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = "_FUNC_(str, sear[, rep]) - Replaces all occurrences of `sear` with `rep`.",
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: > "_FUNC_(str, search[, replace]) - Replaces all occurrences of `search ` with `replace `.",

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 4]
```diff
@@ -340,6 +340,42 @@ case class EndsWith(left: Expression, right: Expression) extends StringPredicate
...
+  extended = """
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]:
```
    Arguments:
      str - a string expression.
      search - a string expression. If `search` is not found in `str`, source-string is returned unchanged.
      replace - a string expression. If `replace` is not specified or is an empty string, nothing replaces
                the string that is removed from `str`.
```

**Comment by gatorsmile** [comment]
LGTM except two minor comments.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 5]
```diff
@@ -340,6 +340,48 @@ case class EndsWith(left: Expression, right: Expression) extends StringPredicate
...
+/**
+ * Replace all occurrences with string.
+ */
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = "_FUNC_(str, search[, replace]) - Replaces all occurrences of `search` with `replace`.",
+  extended = """
+    Arguments:
+      str - a string expression
+      search - a string expression. If `search` is not found in `str`, source-string is returned unchanged.
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: Sorry, `source-string` needs to be updated to `str`.

Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: Good catch. done.
Thanks while it is a long weekend.

**Comment by kiszk** [comment]
Jenkins, retest this please

**Comment by gatorsmile** [comment]
Thanks! Merging to master.
